### ISSUE TITLE
PR evaluation: verbose mode in transaction.get

### ISSUE THREAD

**Issue by gdassori** [kyuupichan/electrumx#385]
Hello,
in bitcoind getrawtransactions supports verbose mode, instead electrum server doesn't: https://github.com/kyuupichan/electrumx/blob/e2b15d7fe117ef3ebe1e91c42772dbe82a8c5c1f/server/daemon.py#L267

Getrawtransaction with verbose mode returns block hash (if any):
"blockhash": "00000000000000000029b786fd3da3c2859a9be4807104d7e80112cdd5a33407",
in addition to the transaction data.

I'm writing a script which uses electrum network to gather info about bitcoin transactions, and the blockhash would be very useful to me (I'm monitoring "transactions" and not a specific address, so address.get_history API is not useful).

Actually, verbose mode response may just be:

```
{ 
  "hex": <hex transaction>
  "blockhash": None or <block hash>
}

```
and not the json representation of the transaction, as bitcoind does. To me is enough and the information is very precious.

I may write a PR and it should be a very low hanging fruit. This issue is to understand if it's welcome.
Thanks !

**Comment by erasmospunk** [comment]
@gdassori keep in mind that not all daemons support this newer API so it you call it without a check, it will break a lot of clients.

**Comment by gdassori** [comment]
@erasmospunk my client can track the protocol version to ensure the new API is available, or possibly switch server.

Above all, as far as I can see, backward compatibility (given a default false value) is granted, which is crucial.

**Comment by erasmospunk** [comment]
I think mislead you with the "daemon" terminology. What I meant is that not all coins support the new getrawtransaction API.

**Comment by gdassori** [comment]
Yes, now I see your point :-) 

Actually I'm not very informed about the alt-currencies ecosystem, but I guess those kind of coins wouldn't benefit from "the blockhash" information (if any exists) in getrawtransaction, and so the relative implementation wouldn't be affected in any way. Do you agree ?

I don't want to advocate a patchwork that is difficult to explain but 
"the API is enabled only since <N> protocol version, and only if the coin backend supports it" 
still sounds reasonable to me.

If design ACK, I will also add a system check (and, obv, a test) to the PR to ensure the API is supported by the coin, before enabling it. 

Actually, default may be False and coins mantainers that wants to be enabled can open a PR, unless is well known the verbose mode is supported (let's say, Litecoin).

**Comment by erasmospunk** [comment]
If the verbose parameter is not present, better not include it in the RPC call at all. Also, I haven't checked all supported coins but it seems the most incompatible ones (peercoin, vcash, decred) require `1` instead of `true`, bitcoin works with `1` as well.

Regarding the return value, it is more useful to return the parsed JSON from the coin daemon so lightweight clients would benefit as well.

@kyuupichan what's your take?

**Comment by kyuupichan** [comment]
@erasmospunk I'm happy to extend the blockchain.transaction.get call to add a new optional parameter, say verbose, that defaults to False.  If true then we can return a dictionary with the raw tx and the block hash.  There is little value in the other fields as presumably the client has the headers already so I think we should only support that one extra field.  It's a shame height isn't provided as that would be useful I suspect - otherwise the client needs a block header hash->height index.  It might be a good idea to add one to ElectrumX.
For daemons that don't support it we should return an RPC error.

**Comment by kyuupichan** [comment]
@gdassori - is this something you plan to work on?  It's not too hard a modification; the main thing is to properly handle the daemon returning an error because it doesn't understand the verbose flag you pass it.

**Comment by gdassori** [comment]
Hi, I wrote a bit here:
https://github.com/kyuupichan/electrumx/compare/master...gdassori:patch-getrawtransaction-verbose

Now I'm testing a bit in my environment, to see how it's going before opening a PR :-)

**Comment by kyuupichan** [comment]
OK some comments - let's not change coins.py; any failure will be correctly passed back to the client.  Cast the boolean to an integer 0 or 1 with int(verbose).  transaction_get_1_0 should not change; I do not wish to change protocol 1.0; it is obsolete.

I guess I'm OK just returning whatever the daemon returns rather than specifying what it is.  

So I think this boils down to just a couple of lines of code changes.  You should update the docs in PROTOCOL.rst specifying this new optional parameter as part of protocol 1.2.  The code will silently accept it for 1.1, which is fine, but we cannot change the spec for that now as it's in the past.  Whilst you're changing the docs, please add a line saying that the result is a JSON null if the transaction is not found.

**Comment by gdassori** [comment]
To avoid misunderstanding: you are suggesting to forward the verbose mode without checking if enabled on the coin, and just handle the error from the daemon? Yeah, it's clean, actually.. perhaps my concern not to send bitcoind calls that could produce errors, is excessive.

Initially I thought I may format the output this way:
```
{
 "hex": <rawtx>,
 "blockhash": <the block hash, if present>,
 "blockheight": <the block height, if present>
}
```
and fetch the block height from a getblockheader call to bitcoind (I still don't know what ElectrumX index by itself, and if there's an available local data repository to fetch this information), but I don't know if this is supported by other coins, and may drive to unexpected errors, so I though just returning roughly the backend response may be the best option. 

Furthermore, I don't know if adding the height is good: I tend to think a well written client implementation should track the headers by itself.

**Comment by kyuupichan** [comment]
Yes. So you only need to change a couple of lines. The error, if any, will be reported already.

**Comment by gdassori** [comment]
@kyuupichan, after playing a bit around the PR, I finally considered that the best solution is the one proposed initially, as now in PR #397.

A deserialization of a transaction that is too large will probably cause network latencies and timeouts and the API may turn be unusable on big transactions (we don't have to forget the hex is returned too).
the crucial information, at the end of the day, is the blockhash, and any decent client should be able to deserialize the transaction by itself.

Please, I'm not native english speaker so, along with the code, feel free to correct also the description :-). Thanks!

**Comment by kyuupichan** [comment]
No it's not an issue; don't worry about it.  Unless you are in the habit of requesting big historical transactions, when you might bump into the server-specific request size limit.  If you are, then you need to persuade an operator to bump up their default, or I might consider doing it in the source.

**Comment by gdassori** [comment]
Ok, as told, I was concerned about hitting the limits. It's because I've read elsewhere about issues (I wouldn't be wrong, I'm kinda new to the repo) about a too long response while listing transactions per address on addresses with a huge history. Thought it may be an issue here too. 
However, the new commit in the PR return the entire deserialization.

**Comment by kyuupichan** [comment]
Implemented in #397
